### PR TITLE
New blocks

### PR DESCRIPTION
**Pull request by rtsisyk**
Add Macro and Micro blocks.

See #828

### CODE REVIEW

File: blockchain/protos/blockchain.proto [hunk 1]
```diff
@@ -42,10 +42,10 @@ message BaseBlockHeader {
     stegos.crypto.VRF random = 6;
 }
 
-message MicroBlockHeader {
+message MacroBlockHeader {
     BaseBlockHeader base = 1;
     stegos.crypto.Fr gamma = 2;
-    int64 monetary_adjustment = 3;
+    int64 block_reward = 3;
     stegos.crypto.Hash inputs_range_hash = 4;
```
Comment (vldm) [inline blockchain/protos/blockchain.proto]: Could block reward be negative?

File: blockchain/src/block.rs [hunk 2]
```diff
@@ -94,53 +100,193 @@ impl Hashable for BaseBlockHeader {
...
+            block_fee,
+            gamma,
+            outputs,
+        };
+
+        MicroBlock::new(base, view_change_proof, coinbase, transactions, pkey)
+    }
+
+    /// Sign block using leader's signature.
+    pub fn sign(&mut self, skey: &secure::SecretKey, pkey: &secure::PublicKey) {
```
Comment (vldm) [inline blockchain/src/block.rs]: Why we need separate method for signing, cant we sign messages in constructors?

File: blockchain/src/blockchain.rs [hunk 3]
```diff
@@ -62,12 +63,12 @@ impl ChainInfo {
     /// Create ChainInfo from micro block.
     /// ## Panics
     /// if view_change is equal to 0
-    pub fn from_micro_block(block: &MicroBlock) -> Self {
-        assert_ne!(block.header.base.view_change, 0);
+    pub fn from_block(base_header: &BaseBlockHeader) -> Self {
```
Comment (vldm) [inline blockchain/src/blockchain.rs]: Chain info is used only for viewchanging in microblock, why we now need constructor from macroblock?

File: blockchain/protos/blockchain.proto (hunk 1 above)
Comment (rtsisyk) [reply blockchain/protos/blockchain.proto]: No. There was idea that slashing should burn money. This idea is not actual anymore.

File: blockchain/src/blockchain.rs (hunk 3 above)
Comment (rtsisyk) [reply blockchain/src/blockchain.rs]: I'm not sure. Probably it is just refactoring bug. Anyway, this method only needs BaseBlockHeader, so requiring entire MicroBlock here is not necessary.

File: blockchain/src/block.rs (hunk 2 above)
Comment (rtsisyk) [reply blockchain/src/block.rs]: I found that in most tests cases we don't need signed MicroBlock. Moreover, we can't sign MacroBlock in constructors. I decided to create a separate method to unify Micro and MacroBlocks. I'm not sure that it was right idea.

File: blockchain/src/blockchain.rs (hunk 3 above)
Comment (vldm) [reply blockchain/src/blockchain.rs]: Requiring MicroBlock here is just make API more strict.

Comment (rtsisyk) [reply blockchain/src/blockchain.rs]: I prefer to don't require data which I don't need. This method only needs view_change, height and previous hash. Why I should force user to pass entire MicroBlock here?

**Comment by rtsisyk** [comment]
Thanks!

**Pull request merged**